File: gpt-3.5-ft-large/template_context_full/exit_data_directive.f90
#ifndef T1
!T1:exit data directive,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER :: errors = 0


        IF (errors .eq. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END
#endif

      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
#ifndef T1
        LOGICAL :: test1
#endif
        failed = .FALSE.
        failcode = 0
#ifndef T1
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
#endif
        CALL EXIT (failcode)
      END PROGRAM

c1      ATOMIC
c2      ATOMIC
      CONTAINS
      SUBROUTINE init_matrices1
        REAL, DIMENSION(LOOPCOUNT, LOOPCOUNT) :: a, b, c
        INTEGER :: x, y

        !$acc enter data create(a(1:LOOPCOUNT, 1:LOOPCOUNT), b(1:LOOPCOUNT, 1:LOOPCOUNT), c(1:LOOPCOUNT, 1:LOOPCOUNT))
        
        SEEDDIM(1) = 1
#       ifdef SEED
        SEEDDIM(1) = SEED
#       endif
        CALL RANDOM_SEED(PUT=SEEDDIM)

        CALL RANDOM_NUMBER(a)
        CALL RANDOM_NUMBER(b)
        CALL RANDOM_NUMBER(c)

        !$acc update device(a(1:LOOPCOUNT, 1:LOOPCOUNT), b(1:LOOPCOUNT,1:LOOPCOUNT), c(1:LOOPCOUNT,1:LOOPCOUNT))
      END SUBROUTINE init_matrices1

c1test  SEQUENTIAL
      INTEGER FUNCTION test1_c1
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        INTEGER :: x, y, errors = 0
        REAL(8), DIMENSION(LOOPCOUNT, LOOPCOUNT) :: a, b, c
        INTEGER :: c1, test1_c1

        !$acc data copyin(a(1:LOOPCOUNT, 1:LOOPCOUNT), b(1:LOOPCOUNT, 1:LOOPCOUNT)) copy(c(1:LOOPCOUNT, 1:LOOPCOUNT))
          DO WHILE (test1_c1 .ne. 0)
            test1_c1 = 0
            !$acc parallel loop reduction(+:c1)
            DO x = 1, LOOPCOUNT
              DO y = 1, LOOPCOUNT
                c(x, y) = c(x, y) + a(x, y) + b(x, y)
              END DO
            END DO
            !$acc parallel
            DO x = 1, LOOPCOUNT
              DO y = 1, LOOPCOUNT
                IF (c(x, y) .lt. a(x, y) .or. c(x, y) .lt. b(x, y)) THEN
                  test1_c1 = 1
                END IF
              END DO
            END DO
          END DO
        !$acc enter data copyout(c(1:LOOPCOUNT, 1:LOOPCOUNT)) reference(a(1:LOOPCOUNT,1:LOOPCOUNT), b(1:LOOPCOUNT, 1:LOOPCOUNT))
        !$acc exit data copyout(c(1:LOOPCOUNT, 1:LOOPCOUNT)) delete(a(1:LOOPCOUNT,1:LOOPCOUNT),b(1:LOOPCOUNT,1:LOOPCOUNT))
        DO x = 1, LOOPCOUNT
          DO y = 1, LOOPCOUNT
            IF (abs(c(x, y)) .gt. PRECISION) THEN
              errors = errors + 1
            END IF
          END DO
        END DO
        test1_c1 = errors
      END

c2test  SEQUENTIAL
      INTEGER FUNCTION test1_c2
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        INTEGER :: x, y, errors = 0
        REAL(8), DIMENSION(LOOPCOUNT, LOOPCOUNT) :: a, b, c
        INTEGER :: c1, test1_c2

        !$acc data copyin(a(1:LOOPCOUNT, 1:LOOPCOUNT), b(1:LOOPCOUNT, 1:LOOPCOUNT)) copy(c(1:LOOPCOUNT, 1:LOOPCOUNT))
          DO WHILE (test1_c2 .ne. 0)
            test1_c2 = 0
            !$acc parallel loop reduction(+:c2)
            DO x = 1, LOOPCOUNT
              DO y = 1, LOOPCOUNT
                c(x, y) = c(x, y) + a(x, y) + b(x, y)
              END DO
            END DO
            !$acc parallel
            DO x = 1, LOOPCOUNT
              DO y = 1, LOOPCOUNT
                IF (c(x, y) .lt. a(x, y) - PRECISION OR c(x, y) .lt. b(x, y) - PRECISION) THEN
                  test1_c2 = 1
                END IF
              END DO
            END DO
          END DO
        !$acc enter data copyout(c(1:LOOPCOUNT, 1:LOOPCOUNT)) reference(a(1:LOOPCOUNT, 1:LOOPCOUNT), b(1:LOOPCOUNT, 1:LOOPCOUNT))
        !$acc exit data copyout(c(1:LOOPCOUNT, 1:LOOPCOUNT)) delete(a(1:LOOPCOUNT,1:LOOPCOUNT),b(1:LOOPCOUNT,1:LOOPCOUNT))
        DO x = 1, LOOPCOUNT
          DO y = 1, LOOPCOUNT
            IF (abs(c(x, y) - (a(x, y) + b(x, y))) .gt. PRECISION) THEN
              errors = errors + 1
            END IF
          END DO
        END DO
        test1_c2 = errors
      END

        SUBROUTINE init_matrices1
          REAL(8), DIMENSION(LOOPCOUNT, LOOPCOUNT) :: a, b, c
          INTEGER :: x, y
          INTEGER :: c1, c2

          !$acc enter data create(a(1:LOOPCOUNT, 1:LOOPCOUNT), b(1:LOOPCOUNT, 1:LOOPCOUNT), c(1:LOOPCOUNT, 1:LOOPCOUNT))
          
          SEEDDIM(1) = 1
#         ifdef SEED
          SEEDDIM(1) = SEED
#         endif
          CALL RANDOM_SEED(PUT=SEEDDIM)

          CALL RANDOM_NUMBER(a)
          CALL RANDOM_NUMBER(b)
          CALL RANDOM_NUMBER(c)

          !$acc update device(a(1:LOOPCOUNT, 1:LOOPCOUNT), b(1:LOOPCOUNT,1:LOOPCOUNT), c(1:LOOPCOUNT,1:LOOPCOUNT))

          c1 = 0
          c2 = 0

          !$acc data exit copyout(c(1:LOOPCOUNT, 1:LOOPCOUNT)) delete(a(1:LOOPCOUNT, 1:LOOPCOUNT), b(1:LOOPCOUNT, 1:LOOPCOUNT))

          DO x = 1, LOOPCOUNT
            DO y = 1, LOOPCOUNT
              c1 = c1 + a(x, y) + b(x, y)
              c2 = c2 + c(x, y)
            END DO
          END DO

          IF (c1 .NE. c2) THEN
            WRITE(*, *) c1
            WRITE(*, *) c2
            CALL EXIT(1)
          END IF
        END SUBROUTINE init_matrices1

        SUBROUTINE init_matrices2
          REAL(8), DIMENSION(LOOPCOUNT, LOOPCOUNT) :: a, b, c
          INTEGER :: x, y

          SEEDDIM(1) = 1
#         ifdef SEED
          SEEDDIM(1) = SEED
#         endif
          CALL RANDOM_SEED(PUT=SEEDDIM)

          CALL RANDOM_NUMBER(a)
          CALL RANDOM_NUMBER(b)
          CALL RANDOM_NUMBER(c)

          !$acc enter data create(c(1:LOOPCOUNT, 1:LOOPCOUNT))
          !$acc enter data create(a(1:LOOPCOUNT, 1:LOOPCOUNT), b(1:LOOPCOUNT, 1:LOOPCOUNT))
          
          !$acc update device(a(1:LOOPCOUNT, 1:LOOPCOUNT), b(1:LOOPCOUNT, 1:LOOPCOUNT), c(1:LOOPCOUNT, 1:LOOPCOUNT))

          !$acc exit data delete(a(1:LOOPCOUNT, 1:LOOPCOUNT), b(1:LOOPCOUNT, 1:LOOPCOUNT)) copyout(c(1:LOOPCOUNT, 1:LOOPCOUNT))
          
        END SUBROUTINE init_matrices2

        INTEGER FUNCTION test1()
          USE OPENACC
          IMPLICIT NONE
          INCLUDE "acc_testsuite.Fh"
          INTEGER :: test1_c1, test1_c2
          INTEGER :: failed
          failed = 0

          !$acc data
          CALL init_matrices1

          IF (test1_c1() .NE. 0) THEN
            failed = 1
          END IF
          IF (test1_c2() .NE. 0) THEN
            failed = 1
          END IF

          !$acc end data

          CALL init_matrices2
          
          CALL EXIT (failed)
        END 

      END PROGRAM

File: bad-tests-v2/exit_data_directive.c
#ifndef T1
//T1:construct-independent,enter-data,V:2.0-2.7
int test1(){
    int err = 0;
    srand(SEED);
    real_t * a = (real_t *)malloc(n * sizeof(real_t));
    real_t * b = (real_t *)malloc(n * sizeof(real_t));

    for (int x = 0; x < n; ++x){
        a[x] = rand() / (real_t)(RAND_MAX / 10);
        b[x] = 0;
    }

    #pragma acc enter data copyin(a[0:n])
    #pragma acc enter data copyin(b[0:n])
    #pragma acc parallel present(a[0:n], b[0:n])
    {
        #pragma acc loop
        for (int x = 0; x < n; ++x){
            b[x] = a[x];
        }
    }
    #pragma acc exit data delete(a[0:n], b[0:n])

    for (int x = 0; x < n; ++x){
        if (fabs(a[x] - b[x]) > PRECISION){
            err += 1;
        }
    }

    return err;
}
#endif

#ifndef T2
//T2:construct-independent,enter-data, V:2.0-2.7
int test2(){
    int err = 0;
    srand(SEED);
    real_t * a = (real_t *)malloc(n * sizeof(real_t));
    real_t * b = (real_t *)malloc(n * sizeof(real_t));

    for (int x = 0; x < n; ++x){
        a[x] = rand() / (real_t)(RAND_MAX / 10);
        b[x] = 0;
    }

    #pragma acc enter data create(a[0:n], b[0:n])
    #pragma acc parallel present(a[0:n], b[0:n])
    {
        #pragma acc loop
        for (int x = 0; x < n; ++x){
            b[x] = a[x];
        }
    }
    #pragma acc exit data copyout(a[0:n], b[0:n])

    for (int x = 0; x < n; ++x){
        if (fabs(a[x] - b[x]) > PRECISION){
            err += 1;
        }
    }

    return err;
}
#endif

#ifndef T3
//T3:construct-independent,enter-data, V:2.0-2.7
int test3(){
    int err = 0;
    srand(SEED);
    real_t * a = (real_t *)malloc(n * sizeof(real_t));
    real_t * b = (real_t *)malloc(n * sizeof(real_t));

    for (int x = 0; x < n; ++x){
        a[x] = rand() / (real_t)(RAND_MAX / 10);
        b[x] = 0;
    }

    #pragma acc enter data create(a[0:n], b[0:n])
    #pragma acc parallel present(a[0:n], b[0:n])
    {
        #pragma acc loop
        for (int x = 0; x < n; ++x){
            b[x] = a[x];
        }
    }
    #pragma acc exit data delete(a[0:n], b[0:n])
    for (int x = 0; x < n; ++x){
        if (fabs(a[x] - b[x]) > PRECISION){
            err += 1;
        }
    }

    return err;
}
#endif

#ifndef T4
//T4:construct-independent,enter-data, V:2.0-2.7
int test4(){
    int err = 0;
    srand(SEED);
    real_t * a = (real_t *)malloc(n * sizeof(real_t));
    real_t * b = (real_t *)malloc(n * sizeof(real_t));
    int * hasDevice = (int *)malloc(sizeof(int));

    hasDevice[0] = 1;
    #pragma acc enter data copyin(hasDevice[0:1])
    hasDevice[0] = 0;

    if (hasDevice[0] == 1){
        for (int x = 0; x < n; ++x){
            a[x] = rand() / (real_t)(RAND_MAX / 10);
            b[x] = 0;
        }

        #pragma acc enter data create(a[0:n], b[0:n])
        #pragma acc parallel present(a[0:n], b[0:n])
        {
            #pragma acc loop
            for (int x = 0; x < n; ++x){
                b[x] = a[x];
            }
        }
        if (fabs(b[0]) < PRECISION){
            err = 1;
        }
        #pragma acc exit data delete(a[0:n], b[0:n])
    }

    return err;
}
#endif

#ifndef T5
//T5:construct-independent,enter-data, V:2.0-2.7
int test5(){
    int err = 0;
    srand(SEED);
    real_t * a = (real_t *)malloc(n * sizeof(real_t));
    real_t * b = (real_t *)malloc(n * sizeof(real_t));
    int * hasDevice = (int *)malloc(sizeof(int));

    hasDevice[0] = 1;
    #pragma acc enter data copyin(hasDevice[0:1])
    hasDevice[0] = 0;

    if (hasDevice[0] == 1){
        for (int x = 0; x < n; ++x){
            a[x] = rand() / (real_t)(RAND_MAX / 10);
            b[x] = 0;
        }

        #pragma acc enter data create(a[0:n], b[0:n])
        #pragma acc parallel present(a[0:n], b[0:n])
        {
            #pragma acc loop
            for (int x = 0; x < n; ++x){
                b[x] = a[x];
            }
        }
        #pragma acc exit data copyout(a[0:n], b[0:n])
        if (fabs(b[0]) > PRECISION){
            err = 1;
        }
        for (int x = 1; x < n; ++x){
            if (fabs(a[x] - b[x]) > PRECISION){
                err = 1;
            }
        }
    }

    return err;
}
#endif

int main(){
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif
#ifndef T2
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test2();
    }
    if (failed != 0){
        failcode = failcode + (1 << 1);
    }
#endif
#ifndef T3
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test3();
    }
    if (failed != 0){
        failcode = failcode + (1 << 2);
    }
#endif
#ifndef T4
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test4();
    }
    if (failed != 0){
        failcode = failcode + (1 << 3);
    }
#endif
#ifndef T5
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test5();
    }
    if (failed != 0){
        failcode = failcode + (1 << 4);
    }
#endif
    return failcode;
}